rfc/cart-api-v1.md

```diff
@@ -0,0 +1,251 @@
+```
+
+Here is the sample request.
+
+```
+GET /resources/carts/default
+Authorization: Bearer 261e1aea-82c9-4e38-a879-e2e02ef236a5
+```
+
+Here is the sample response.
+
+```javascript
+{
+    id: "6b801de6-11d4-46c0-aed1-ca2110aa7d3c"
+    cart_name: "",
+    default: true,
+    items: [...]
+}
+```
+
+#### `PUT /resources/carts/{id}`
+
+Here is the request body.
+
+```javascript
+{
+    cart_name: str // Required
+}
+```
+
+and the response is basically the JSON representation of `Cart`.
+
+#### `DELETE /resources/carts/{id}`
+
+* The endpoint will delete the target cart and respond with **HTTP 200** only if the cart belongs to the user making the request.
+  * Otherwise, the endpoint will respond with **HTTP 403**.
+* The endpoint will respond **HTTP 404** if the cart is not found.
+
+### CRUD APIs for `CartItem`
+
+| Method | Path | Description |
+| --- | --- | --- |
+| `POST` | `/resources/carts/{cart_id}/items/` | Create a new cart item |
+| `DELETE` | `/resources/carts/{cart_id}/items/{id}` | Delete a single cart item |
+
+#### `POST /resources/carts/{cart_id}/items/`
+
+| Param Type | Param Name | Association |
+| --- | --- | --- |
+| URL Path | `cart_id` | `Cart.id` |
+
+* The request body will be similar to `CartItem` without `cart_id`.
+* The endpoint will also respond **HTTP 404** if the corresponding cart is not found.
+
+> May need a new field to store SHA1 hash on the item data for quick check.
+
+#### `DELETE /resources/carts/{cart_id}/items/{item_id}`
+
+| Param Type | Param Name | Association |
+| --- | --- | --- |
+| URL Path | `cart_id` | `Cart.id` |
+| URL Path | `item_id` | `CartItem.id` |
+
+* The endpoint will also respond **HTTP 404** if the corresponding cart is not found.
+
+### Other APIs
+
+#### `POST /resources/carts/{cart_id}/items/query`
+
+Save the cart items in batch by query.
+
+| Param Type | Param Name | Association |
+| --- | --- | --- |
+| URL Path | `cart_id` | `Cart.id` |
+
+TL;DR:
+* Respond **HTTP 202** on successfully accepting a request.
+* Use `POST /resources/cart-item-batch-jobs/{job_id}` to check the async job status.
+
+The response body for HTTP 202 will be `CartItemBatchJob`.
+
+##### Technical Considerations
+
+As this operation may require more than 30 seconds to run, the batch operation is divided into three major parts: processing requests, job updates, and job reports.
+
+> See [the overall diagram describing the process](https://github.com/DataBiosphere/azul/raw/rfc/cart-api/rfc/DNAstack%20-%20General%20-%20Batch%20Cart%20Item%20Creation%20API%20Diagram.png).
+
+###### Processing requests
+
+> See "Creation Request" in the diagram.
+
+When the endpoint `λ1` receives a request with filters, the endpoint will delegate the query to `λ2` via **the request queue**. At this point, the client will not be able to figure out whether the new cart items based on the query are in the cart. The client will have to track the progress by sending request to `λ5` for an update.
```

Comment:
Using λ as identifiers is kind of hard to understand.  Should note that λ1 is the `/resources/carts/{cart_id}/items/query` endpoint which returns a request id that is sent to λ5 which is the `/resources/cart-item-batch-jobs/{job_id}` endpoint